File: Java/dynamicprogramming/MinimumCoins.java
package dynamicprogramming;

import java.util.Arrays;

public class MinimumCoins {
  /**
   * Time Complexity => O(length of coins array * target)
   * <p>
   * Space Complexity => O(target)
   * 
   * @param coins  Array of coins
   * @param target Target amount
   * @return Minimum coins required to get the target amount
   */
  public int minCoins(int[] coins, int target) {
    int[] dp = new int[target + 1];
    Arrays.fill(dp, Integer.MAX_VALUE);
    dp[0] = 0;
    for (int i = 1; i <= target; i++) {
      for (int coin : coins) {
        if (coin <= i) {
          int remTarget = 1 + dp[i - coin];
          if (remTarget != Integer.MAX_VALUE)
            dp[i] = Math.min(dp[i], remTarget);
        }
      }
    }
    return dp[target];
  }

  public static void main(String[] args) {
    MinimumCoins ob = new MinimumCoins();
    System.out.println(ob.minCoins(new int[] { 10, 5, 25 }, 30));
  }
}

File: C++/DynamicProgramming/MinimumCoins.cpp
#include <bits/stdc++.h>
using namespace std;

int minCoins(int coins[], int n, int target)
{
  vector<int> dp(target + 1, INT16_MAX);
  dp[0] = 0;
  for (int i = 1; i <= target; i++)
  {
    for (int j = 0; j < n; j++)
    {
      if (coins[j] <= i)
      {
        int remTarget = 1 + dp[i - coins[j]];
        if (remTarget != INT32_MAX)
          dp[i] = min(dp[i], remTarget);
      }
    }
  }
  return dp[target];
}

int main()
{
  int coins[] = {10, 25, 5};
  int n = sizeof(coins) / sizeof(coins[0]);
  cout << minCoins(coins, n, 30) << endl;
}